### PR TITLE
Add cursor operations to the `StateMachine` tests

### PR DESCRIPTION
**Pull request by jorisdral**

### CODE REVIEW

File: test/Database/LSMTree/Model/Normal.hs [hunk 1]
```diff
@@ -225,3 +229,36 @@ duplicate ::
...
   -> Table k v blob
 duplicate = id
+
+{-------------------------------------------------------------------------------
+  Cursor
+-------------------------------------------------------------------------------}
+
+type Cursor :: Type -> Type -> Type -> Type
+data Cursor k v blob = Cursor
+    { _cursorValues :: [(RawBytes, (RawBytes, Maybe (BlobRef blob)))]
```
Comment (mheinzel) [inline test/Database/LSMTree/Model/Normal.hs:239]: It's kind of clear from the usage of `Map.toList` already, but might be worth adding a short note that these entries are already resolved, or more generally that the first element of the tuple is the key and entries are sorted+unique by key.

**Review by mheinzel — APPROVED** [review]
This does a bit more than just adding cursor support to the tests, but looks good!

File: src/Database/LSMTree/Common.hs [hunk 2]
```diff
@@ -220,21 +220,23 @@ listSnapshots (Internal.Session' sesh) = Internal.listSnapshots sesh
 --
 -- Though blob references are handle-like, they /do not/ keep files open. As
 -- such, when a blob reference is returned by a lookup, modifying the
--- corresponding table handle (or session) /may/ cause the blob reference to be
--- invalidated (i.e.., the blob has gone missing because the blob file was
+-- corresponding table handle, cursor, or session /may/ cause the blob reference
```
Comment (dcoutts) [inline src/Database/LSMTree/Common.hs:223]: I don't follow. Cursors cannot be modified.

File: src/Database/LSMTree/Common.hs [hunk 3]
```diff
@@ -220,21 +220,23 @@ listSnapshots (Internal.Session' sesh) = Internal.listSnapshots sesh
...
+-- * Closing cursors
 -- * Closing sessions
 --
 -- An invalidated blob reference will throw an exception when used to look up a
--- blob. Note that table operations such as snapshotting and duplication do
--- /not/ invalidate blob references. These operations do not modify the logical
--- contents or state of an existing table.
+-- blob. Note that operations such as snapshotting, duplication and cursor reads
+-- do /not/ invalidate blob references. These operations do not modify the
+-- logical contents or state of a table.
```
Comment (dcoutts) [inline src/Database/LSMTree/Common.hs:235]: :+1:

File: src-extras/Database/LSMTree/Extras/NoThunks.hs [hunk 4]
```diff
@@ -348,8 +350,21 @@ deriving anyclass instance NoThunks RefCount
...
-deriving stock instance Generic (MutableHeap s a)
-deriving anyclass instance NoThunks a => NoThunks (MutableHeap s a)
+instance (NoThunks a, Typeable s, Typeable a) => NoThunks (MutableHeap s a) where
+  showTypeOf (p :: Proxy (MutableHeap s a)) = show $ typeRep p
+  wNoThunks ctx
+    (MH (a :: PrimVar s Int) (b :: SmallMutableArray s a))
+    = allNoThunks [
+          noThunks ctx a
+          -- the small array may contain bogus/undefined placeholder values
+          -- after the first @n@ elements in the heap
```
Comment (dcoutts) [inline src-extras/Database/LSMTree/Extras/NoThunks.hs:360]: Good catch.

**Review by dcoutts — APPROVED** [review]
Looks good.

File: src/Database/LSMTree/Common.hs (hunk 2 above)
Comment (jorisdral) [reply src/Database/LSMTree/Common.hs:223]: Technically, closing cursor counts as modification, that's also why session close is listed as something that can invalidate a blob reference

**Pull request merged**